apps/storeutl.c

```diff
@@ -0,0 +1,192 @@
+    {"out", OPT_OUT, '>', "Output file - default stdout"},
+    {"passin", OPT_PASSIN, 's', "Input file pass phrase source"},
+    {"text", OPT_TEXT, '-', "Print a text form of the objects"},
+    {"noout", OPT_NOOUT, '-', "No PEM output, just status"},
+#ifndef OPENSSL_NO_ENGINE
+    {"engine", OPT_ENGINE, 's', "Use engine, possibly a hardware device"},
+#endif
+    {NULL}
+};
+
+int storeutl_main(int argc, char *argv[])
+{
+    OSSLSTORE_CTX *store_ctx = NULL;
+    int ret = 1, noout = 0, text = 0, items = 0;
+    char *outfile = NULL, *passin = NULL, *passinarg = NULL;
+    BIO *out = NULL;
+    ENGINE *e = NULL;
+    OPTION_CHOICE o;
+    char *prog = opt_init(argc, argv, storeutl_options);
+    PW_CB_DATA pw_cb_data;
+
+    while ((o = opt_next()) != OPT_EOF) {
+        switch (o) {
+        case OPT_EOF:
+        case OPT_ERR:
+ opthelp:
+            BIO_printf(bio_err, "%s: Use -help for summary.\n", prog);
+            goto end;
+        case OPT_HELP:
+            opt_help(storeutl_options);
+            ret = 0;
+            goto end;
+        case OPT_OUT:
+            outfile = opt_arg();
+            break;
+        case OPT_PASSIN:
+            passinarg = opt_arg();
+            break;
+        case OPT_NOOUT:
+            noout = 1;
+            break;
+        case OPT_TEXT:
+            text = 1;
+            break;
+        case OPT_ENGINE:
+            e = setup_engine(opt_arg(), 0);
+            break;
+        }
+    }
+    argc = opt_num_rest();
+    argv = opt_rest();
+
+    if (argc == 0) {
+        BIO_printf(bio_err, "%s: No URI given, nothing to do...\n", prog);
+        goto opthelp;
+    }
+    if (argc > 1) {
+        BIO_printf(bio_err, "%s: Unknown extra parameters after URI\n", prog);
+        goto opthelp;
+    }
+
+    if (!app_passwd(passinarg, NULL, &passin, NULL)) {
+        BIO_printf(bio_err, "Error getting passwords\n");
+        goto end;
+    }
+    pw_cb_data.password = passin;
+    pw_cb_data.prompt_info = argv[0];
+
+    out = bio_open_default(outfile, 'w', FORMAT_TEXT);
+    if (out == NULL)
+        goto end;
+
+    if ((store_ctx = OSSLSTORE_open(argv[0], get_ui_method(), &pw_cb_data, NULL,
+                                   NULL)) == NULL) {
+        BIO_printf(bio_err, "Couldn't open file or uri %s\n", argv[0]);
+        ERR_print_errors(bio_err);
+        goto end;
+    }
+
+    /* From here on, we count errors, and we'll return the count at the end */
+    ret = 0;
+
+    for (;;) {
+        OSSLSTORE_INFO *info = OSSLSTORE_load(store_ctx);
+        const char *infostr = NULL;
+
+        if (info == NULL) {
+            if (OSSLSTORE_eof(store_ctx))
+                break;
+
+            if (OSSLSTORE_error(store_ctx)) {
+                ERR_print_errors(bio_err);
+                ret++;
+                break;
+            }
+        }
+
+        /* If info is NULL at this point, then the loader is badly coded */
+        OPENSSL_assert(info != NULL);
```

Comment:
Mm...  considering this is an app, and that the correct response is to break out, I wasn't too concerned.  But ok, it's probably a nice thing to allow the store to close either way, so I'll just print an error and break out of the loop.